### ISSUE TITLE
id attribute should be allowed on all CellML elements

### ISSUE THREAD

**Issue by nickerso** [cellml/libcellml#165]
Just realised while reviewing #160 that the parser will give an error for some elements with an `id` attribute. Since this is an issue for the parser in general, I don't think it should hold up the #160 but rather addressed as a separate issue.

**Comment by hsorby** [comment]
I think we have so far hand waved the id problem, some elements like connections are going to be really hard to put an id on as they don't exist as an object in the object model.

**Comment by hsorby** [comment]
As part of being able to set ids on everything we will need to change the API to incorporate this.  One option is to add API on the ComponentEntity class:

```
ComponentEntity::setEncapsulationId();
ComponentEntity::getEncapsulationId();
```

This would allow us to identify elements under the encapsulation element.

**Comment by hsorby** [comment]
For connections, we could add to the creating an equivalence API:

```
void addEquivalence(const VariablePtr &variable1, const VariablePtr &variable2, const std::string &mappingId="", const std::string &connectionId="");
std::string getEquivalenceMappingId(size_t index);
std::string getEquivalenceConnectionId(size_t index);
```

**Comment by agarny** [comment]
Hmm, not sure I like the way this is / might be going. I find it awkward.

I would personally like to have objects for every single CellML element (thus making it easy and "natural" to set/get an `id` attribute, among other things), as well as the different methods that we already have and which purpose is to make the API easier to use from a developer viewpoint.

**Comment by hsorby** [comment]
Some of this is a little awkward but we want to model the domain we are working in and not the underlying XML.  These little bits of awkwardness I don't think will be important as I don't think too many people will be wanting to annotate the encapsulation or the connection itself.

**Comment by nickerso** [comment]
Just to comment that I agree with Hugh's comment above, as long as we can satisfy these edge cases in the XML serialisation with the current data model, I'd prefer not to be changing it too drastically until we've thoroughly tested it in the way most application developers will be using it.

**Comment by MichaelClerx** [comment]
OK! Let's see how it works out :-)

**Comment by agarny** [comment]
Ugly, but I appreciate what you guys are saying. So, as @MichaelClerx would say: "Let's see how it works out :-)" :)

**Comment by nickerso** [comment]
closed by #227
